Fix override for __iter__

hameerabbasi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54702 Fix override for __iter__**

This fixes subclassing for __iter__ so that it returns an iterator over
subclasses properly instead of Tensor.

Differential Revision: [D27352563](https://our.internmc.facebook.com/intern/diff/D27352563)

hameerabbasi committed:
Fix override for __iter__

This fixes subclassing for __iter__ so that it returns an iterator over
subclasses properly instead of Tensor.

[ghstack-poisoned]

hameerabbasi committed:
Update on "Fix override for __iter__"


This fixes subclassing for __iter__ so that it returns an iterator over
subclasses properly instead of Tensor.

[ghstack-poisoned]

ezyang commented on torch/_tensor.py:
```diff
@@ -585,8 +585,6 @@ def __iter__(self):
         # indexes of hiddens[0] before hiddens[1], while the generator
         # map will interleave them.)
-        if has_torch_function_unary(self):
-            return handle_torch_function(Tensor.__iter__, (self,), self)
```
Probably good to have a comment here saying that torch function handling is intentionally omitted

hameerabbasi replied:
Done in https://github.com/pytorch/pytorch/pull/54702/commits/e6693244022b495a42020c4ee3b00424edbca8e7.

hameerabbasi committed:
Update on "Fix override for __iter__"


This fixes subclassing for __iter__ so that it returns an iterator over
subclasses properly instead of Tensor.

Differential Revision: [D27352563](https://our.internmc.facebook.com/intern/diff/D27352563)

[ghstack-poisoned]
